src/dotnet/commands/dotnet-add-p2p/Program.cs

```diff
@@ -0,0 +1,232 @@
+            CommandOption referenceOption = app.Option("-r|--reference <REFERENCE>", "Add project to project <REFERENCE> to <PROJECT>", CommandOptionType.MultipleValue);
+
+
+            app.OnExecute(() => {
+                ProjectRootElement project = projectArgument.Value != null ?
+                                                GetProjectFromFileOrThrow(projectArgument.Value) :
+                                                GetProjectFromCurrentDirectoryOrThrow();
+
+                if (referenceOption.Values.Count == 0)
+                {
+                    throw new GracefulException("You must specify at least one reference to add.");
+                }
+
+                AddProjectToProjectReference(project, frameworkOption.Value(), referenceOption.Values);
+
+                project.Save();
+
+                return 0;
+            });
+
+            try
+            {
+                return app.Execute(args);
+            }
+            catch (GracefulException e)
+            {
+                Reporter.Error.WriteLine(e.Message);
+                return 1;
+            }
+        }
+
+        // There is ProjectRootElement.TryOpen but it does not work as expected
+        // I.e. it returns null for some valid projects
+        public static ProjectRootElement TryOpenProject(string filename)
+        {
+            try
+            {
+                return ProjectRootElement.Open(filename);
+            }
+            catch (Microsoft.Build.Exceptions.InvalidProjectFileException)
+            {
+                return null;
+            }
+        }
+
+        public static ProjectRootElement GetProjectFromFileOrThrow(string filename)
+        {
+            if (!File.Exists(filename))
+            {
+                throw new GracefulException($"Provided project `{filename}` does not exist.");
+            }
+
+            var project = TryOpenProject(filename);
+            if (project == null)
+            {
+                throw new GracefulException($"Invalid MSBuild project `{filename}`.");
+            }
+
+            return project;
+        }
+
+        public static ProjectRootElement GetProjectFromCurrentDirectoryOrThrow()
+        {
+            DirectoryInfo currDir = new DirectoryInfo(Directory.GetCurrentDirectory());
+            FileInfo[] files = currDir.GetFiles("*proj");
+            if (files.Length == 0)
+            {
+                throw new GracefulException("Could not find any MSBuild project in the current directory.");
+            }
+
+            if (files.Length > 1)
+            {
+                throw new GracefulException("Found more than one MSBuild project in the current directory. Please specify which one to use.");
+            }
+
+            FileInfo projectFile = files.First();
+
+            if (!projectFile.Exists)
```

Comment:
True but the error messages are different when the file is invalid or not present. I.e.:
- when project is passed but is invalid you expect: "... invalid project..."
- when project is passes but doesn't exist you expect: "... project file doesn't exist..."
- when project is not passed, found and invalid you still expect: "...could not find a project..."
- when project is not passed, found but was being edited by other process and actually doesn't exist on disk anymore you expect: "...could not find a project..."
- when found more than one project you expect: "found more than one"

as you see the error messages are completely different so there is not much common in this logic. IMO separate